client/network/src/state_request_handler.rs

```diff
@@ -0,0 +1,247 @@
+// Copyright 2020 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Helper for handling (i.e. answering) state requests from a remote peer via the
+//! [`crate::request_responses::RequestResponsesBehaviour`].
+
+use codec::{Decode};
+use crate::chain::{Client, StorageKey};
+use crate::config::ProtocolId;
+use crate::request_responses::{IncomingRequest, OutgoingResponse, ProtocolConfig};
+use crate::schema::v1::{StateResponse, StateRequest};
+use crate::{PeerId, ReputationChange};
+use futures::channel::{mpsc, oneshot};
+use futures::stream::StreamExt;
+use log::debug;
+use lru::LruCache;
+use prost::Message;
+use sp_runtime::generic::BlockId;
+use sp_runtime::traits::Block as BlockT;
+use std::sync::Arc;
+use std::time::Duration;
+use std::hash::{Hasher, Hash};
+
+const LOG_TARGET: &str = "sync";
+const MAX_RESPONSE_BYTES: usize = 2 * 1024 * 1024; // Actual reponse may be bigger.
```

Comment:
In case we touch the runtime with 'runtime size > 2 megs' then one chunk contains runtime only and next chunk also contains runtime only because we restart next chunk on this single key, then infiniloop ?
That sounds actually a bit tricky, first idea would be to only stop when we do not have a key that starts with previous key.

Reply:
The response does not actually start with the `start` key, but with the one following it. I should add this to the documentation somewhere.

Reply:
Oh yes, it's working, no real need for doc. (the fact that we always got a key even if proof size is above threshold is what matters).